data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/PeerForwarder.java

```diff
@@ -85,69 +93,70 @@ public PeerForwarder(final PluginSetting pluginSetting) {
+            for (final Span span : entry.getValue()) {
+                final ResourceSpans rs;
+                try {
+                    rs = oTelProtoEncoder.convertToResourceSpans(span);
+                    currRequestBuilder.addResourceSpans(rs);
+                    currBatchSpans.add(span);
+                } catch (UnsupportedEncodingException | DecoderException e) {
+                    LOG.error("failed to encode span with spanId: {} into opentelemetry-protobuf, span will be exported locally.",
+                            span.getSpanId(), e);
+                    recordsToProcessLocally.add(new Record<>(span));
+                }
+                if (currBatchSpans.size() >= maxNumSpansPerRequest) {
                     final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
-                    if (isLocalClient(client)) {
-                        recordsToProcessLocally.add(new Record<>(currRequest));
-                    } else {
-                        forwardedRequestFutures.add(processRequest(client, currRequest));
-                    }
+                    forwardedRequestFuturesToSpans.put(processRequest(client, currRequest), currBatchSpans);
                     currRequestBuilder = ExportTraceServiceRequest.newBuilder();
-                    currSpansCount = 0;
+                    currBatchSpans = new ArrayList<>();
                 }
-                currRequestBuilder.addResourceSpans(rs);
-                currSpansCount += rsSize;
             }
             // Dealing with the last batch request
-            if (currSpansCount > 0) {
+            if (currBatchSpans.size() > 0) {
                 final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
-                if (client == null) {
-                    recordsToProcessLocally.add(new Record<>(currRequest));
-                } else {
-                    forwardedRequestFutures.add(processRequest(client, currRequest));
-                }
+                forwardedRequestFuturesToSpans.put(processRequest(client, currRequest), currBatchSpans);
             }
         }
 
-        for (final CompletableFuture<Record> future : forwardedRequestFutures) {
+        for (final Map.Entry<CompletableFuture<ExportTraceServiceRequest>, List<Span>> entry : forwardedRequestFuturesToSpans.entrySet()) {
             try {
-                final Record record = future.get();
-                if (record != null) {
-                    recordsToProcessLocally.add(record);
+                final CompletableFuture<ExportTraceServiceRequest> future = entry.getKey();
+                final ExportTraceServiceRequest request = future.get();
+                if (request != null) {
+                    final List<Span> spansFailedToForward = forwardedRequestFuturesToSpans.get(future);
+                    recordsToProcessLocally.addAll(spansFailedToForward.stream().map(Record::new).collect(Collectors.toList()));
```

Comment:
I like that you are still attempting to process traces even if there are failures. What about the catch statement below. Should we include the spans which failed to send because of an exception?

Reply:
Great catch! It seems to be a miss before. Will doublecheck